libminifi/src/utils/Id.cpp

```diff
@@ -193,8 +189,8 @@ IdGenerator::~IdGenerator() = default;
 
 uint64_t IdGenerator::getDeviceSegmentFromString(const std::string& str, int numBits) const {
   uint64_t deviceSegment = 0;
-  for (size_t i = 0; i < str.length(); i++) {
-    unsigned char c = toupper(str[i]);
+  for (auto ch : str) {
+    unsigned char c = toupper(ch);
```

Comment:
The argument of `toupper` needs to fit into an `unsigned char`.  We have `static_cast<unsigned char>` elsewhere, but a `gsl::narrow` may be better.

Reply:
This is a hexadecimal digit coming from the `uid.minifi.device.segment` property in the config file, so terminating (not starting up) would be OK, but if we don't want to change the behavior in this refactoring PR, that's fine, too.